htlcswitch: fix error misclassification for on-chain resolutions

calvinrzachman opened the pull request:
## Change Description
This PR demonstrates a subtle bug in the `htlcswitch` that causes an incorrect error to be returned for on-chain HTLC resolutions. On the introduction of the `switchrpc` service in #9489, this bug can be escalated, leading to an ambiguous scenario where the Switch reports a payment result which has no forwarding error, pre-image, or encrypted error for the RPC client.

### Problem

When the `contractcourt` resolves a [dust HTLC](https://github.com/lightningnetwork/lnd/blob/v0.20.0-beta/contractcourt/channel_arbitrator.go#L1007) on-chain, it [sends a `ResolutionMsg`](https://github.com/lightningnetwork/lnd/blob/v0.20.0-beta/contractcourt/channel_arbitrator.go#L3381) to the switch that results in an `UpdateFailHTLC` with a `nil` error `Reason`. 

The core issue stems from a a subtle artifact in the `networkResult` store's persistence logic:
1.  A `nil` `htlc.Reason` slice is serialized to disk as a zero-length byte array.
2.  Upon deserialization, this is read back as a **_non-nil, empty_** byte slice (`[]byte{}`).

This `nil` -> `[]byte{}` transformation causes the [check](https://github.com/lightningnetwork/lnd/blob/v0.20.0-beta/htlcswitch/switch.go#L1093) for on-chain resolutions to fail. As a result, this code path, which is intended to return `FailPermanentChannelFailure` to signal the channel is being resolved on-chain, is never executed for results read from disk.

Instead, the code falls through to the `default` case. The Switch attempts to decrypt the empty slice, which fails, causing `parseFailedPayment` to [return `ErrUnreadableFailureMessage`](https://github.com/lightningnetwork/lnd/blob/v0.20.0-beta/htlcswitch/switch.go#L1116).

The impact on `master` is a **misclassification of the error**. The `ChannelRouter` receives a generic, imprecise error instead of the correct `FailPermanentChannelFailure`, hindering its ability to accurately understand the payment's terminal state.

### Solution

The fix is to make the check in `parseFailedPayment` robust to the serialization artifact by checking the length of the reason instead of whether it's `nil`.

```diff
--- a/htlcswitch/switch.go
+++ b/htlcswitch/switch.go
-       case isResolution && htlc.Reason == nil:
+       case isResolution && len(htlc.Reason) == 0:
```

This change ensures that on-chain resolutions are correctly identified and processed, whether the `Reason` is `nil` (in-memory) or an empty slice (post-deserialization). This makes the intended code path reachable and allows the switch to return the precise error to the `ChannelRouter`.

## Steps to Test
- Run this to demo the `nil` -> `[]byte{}` transformation during a disk round-trip: 
  - `go test -v -timeout 30s --tags dev -run ^TestNetworkResultSerialization$ github.com/lightningnetwork/lnd/htlcswitch`
- Then run this more realistic `TestExtractResult`, which simulates the disk I/O, both with and _without_ the final commit which provides the fix: 
  - `go test -v -timeout 30s -run ^TestExtractResult$ github.com/lightningnetwork/lnd/htlcswitch`

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

calvinrzachman committed:
htlcswitch: fix on-chain resolution error misclassification

This commit fixes a bug in `parseFailedPayment` where an on-chain
resolution with an empty reason was not being correctly identified
after being deserialized from disk by SubscribeResult or
GetResult. This impacts callers waiting for the final settle/fail
result on an htlc attempt via GetAttemptResult.

The lnwire codec transforms a `nil` reason into a non-nil, empty
byte slice and bypasses the existing check for `htlc.Reason == nil`.
Thankfully this does not crash the daemon. Rather, it just leads to
the caller (in our case, the ChannelRouter) not actually receiving
the expected FailPermanentChannelFailure link error.

ziggie1984 commented on htlcswitch/switch_test.go:
```diff
@@ -5554,3 +5555,177 @@ func testSwitchAliasInterceptFail(t *testing.T, zeroConf bool) {
+				isResolution: true,
+			},
+			expectedResult: &PaymentResult{
+				Error: NewDetailedLinkError(
```
so this test fails with the old nil check ?

ziggie1984 commented:
not super high prio because the local router still flags it as a failed payment but I think we can add the fix if you polish this PR.

calvinrzachman committed:
htlcswitch: add test to document on-chain resolution bug

This commit adds a new test, TestOnChainResolutionFailure,
which demonstrates a error misclassification in the htlcswitch.

The test shows that when an on-chain resolution with a nil/empty
error reason is read from the result store, it is misclassified by
the parseFailedPayment function. On master, this leads to an
incorrect error being returned to the ChannelRouter by the
GetAttemptResult function.

This test documents the current (buggy) on-chain resolution behavior
and is corrected in the following commit.

lightninglabs-deploy commented:
@calvinrzachman, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@calvinrzachman, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@calvinrzachman, remember to re-request review from reviewers when ready

litbot-9000 commented:
@bitromortac: review reminder
@calvinrzachman, remember to re-request review from reviewers when ready

bitromortac reviewed (approved):
LGTM nice catch :zap:

ziggie1984 reviewed (approved):
Thank you for the fix, can you fix the linter pls before merge
